[8주차] 김시온

enohs opened the pull request.

enohs committed:
6주차 과제 중간 제출입니다! 세션 후에 조금 더 마무리하고 올리겠습니다.

enohs committed:
[6주차] 김시온 학습 PR 제출합니다.

enohs committed:
Add files via upload

enohs committed:
Delete SpringFlow.png

enohs committed:
Add files via upload

Spring 흐름도 첨부

enohs committed:
Update [6주차]김시온.md

[6주차] 김시온 학습 PR 제출합니다.

KoSeonJe commented on week06/[6주차]김시온.md:
```diff
@@ -0,0 +1,228 @@
+
+# 필수
+
+## 프레임워크란?
```
라이브러리와 차이점을 알아보면 이해하기 쉽습니다!

KoSeonJe commented on week06/[6주차]김시온.md:
```diff
@@ -0,0 +1,228 @@
+
+---
+
+## AOP란?
```
Spring MVC 외부에서 지원하는 Filter,
Spring MVC 내부에서 지원하는 Interceptor를 함께 할아보면 좋을 것 같습니다

KoSeonJe commented on week06/[6주차]김시온.md:
```diff
@@ -0,0 +1,228 @@
+## Spring에서 Bean 관리 방법
+빈은 인스턴스화된 객체를 의미하며 스프링 컨테이너에 등록된 객체를 스프링 빈이라고 한다. new 키워드 대신 사용하는 것이라고 볼 수 있다고 한다. 스프링에서 빈을 사용하는 이유는 스프링 간 객체가 의존관계를 관리하도록 하는 것에 있다고 한다. 객체가 의존관계를 등록하면 스프링 컨테이너에서 해당하는 빈을 찾고 그 빈과 의존성을 만든다.
+   
+빈을 등록할 때에는 xml에 직접 등록하거나 @Bean을 사용하거나, @Componenet, @Controller, @Service, @Repository를 사용하는 방법이 있다.
```
빈 등록 방법에는 두 가지가 있는데, 자세히 알아보면 좋을 것 같습니다

KoSeonJe commented on week06/[6주차]김시온.md:
```diff
@@ -0,0 +1,228 @@
+## Layered Architecture
+소프트웨어 아키텍처는 소프트웨어 시스템의 전체적인 구조와 구성요소들 간의 상호작용을 설계하는 것을 말한다. 건물의 뼈대를 잡는 것과 같은 과정인 것이다. 소프트웨어 아키텍처는 시스템의 전체적인 동작을 결정하고, 시스템의 품질과 성능에 직접적으로 영향을 준다. 선택을 할 때에는 요구사항, 복잡성, 기술 스택, 경험, 비용 및 시간 등의 요소들을 고려하여 적절한 패턴을 선택한다. 그 중 가장 기본적인, 표준인 아키텍처가 레이어드 아키텍처이다.
+   
+레이어드 아키텍처는 소프트웨어 시스템을 관심사 별로 여러 개의 계층으로 분리한 아키텍처를 말한다. 이때 관심사에 따라 계층을 나누는 것을 계층화라고 한다. 각 계층은 관심사가 분리되어 어플리케이션 내에서 특정 역할과 책임이 있고, 각자 자신의 역할에만 집중한다. 서로 추상화된 인터페이스로만 소통을 하며, 이 소통은 자신보다 상위인 계층이나 인접하지 않은 계층에는 요청을하지 못하고 인접한 하위 계층에 요청을 보내는 방식으로 진행된다. 여기서 주목할 것은 단발성 의존성이다. 하위 계층은 상위 계층을 몰라야 하며 하위 계층은 인터페이스만 제공하여 요청만 받는다.
```
하위 계층은 상위 계층을 몰라야 한다는 표현보다, 의존하지 않아야 한다는 표현이 더 적절할 것 같습니다.

글로 풀어 언급해주셨지만, 한마디로 표현하면 상위 -> 하위로만 의존하는 아래 방향으로만 의존하는 관계라고 이해하면 될 것 같습니다

KoSeonJe commented on week06/[6주차]김시온.md:
```diff
@@ -0,0 +1,228 @@
+   
+항상 계층이 닫혀있기만 한 것은 아니다. 특정 기능에서만 사용하기 위한 계층을 따로 만들었을 때 요청이 이 계층을 거치지 않고 바로 다음 계층으로 이동하도록 하는 것이 "OPEN"이다. 레이어 아키텍처에서는 계층 구조와 요청 흐름을 문서화하여 어떤 계층이 열려있고 닫혀있는지를 나타내고 그 이유를 적절히 전달해야 한다.
+   
+레이어드 아키텍처는 관심사별로 계층이 분리되어 있어 코드의 재사용성과 유지보수성을 향상시키며 각 계층이 독립적으로 변화 가능하다. 또한 테스트도 독립적으로 가능하다는 특징이 있다. 다만 계층간 통신을 통해 동작ㅎ하여 데이터의 전달 및 변환 과정에서 오버헤드가 발생하기도 한다. 이는 계층이 많아질 수록 더 증가한다. 그리고 통신을 위한 인터페이스와 로직을 추가해야하기 때문에 복잡성이 증가한다는 단점이 있다.
```
사소하지만 오타가 있습니다 ㅎ
동작ㅎ하여 -> 동작하여

Hoya324 commented on week06/[6주차]김시온.md:
```diff
@@ -0,0 +1,228 @@
+- View
+  - 개발자가 직접 개발하는 컴포넌트로, 보통 UI 화면을 의미한다. 
+
+Spring MVC의 전체적인 동작원리는 다음과 같다.
+1. 클라이언트가 요청을 전송하면 DispatcherServlet이라는 클래스에 요청 전달
+2. DispatcherServlet은 클라이언트의 요청을 처리할 Controller에 대한 검색을 HandlerMapping 인터페이스에게 요청
+3. HandlerMapping은 클라이언트 요청과 매핑되는 핸들러 객체를 다시 DispatcherServlet에게 리턴함 (핸들러 객체는 해당 핸들러의 Handler 메소드 정보를 포함하고 있음. Handler 메소드는 Controller 클래스 안에 구현된 요청 처리 메소드를 의미함)
+4. 요청을 처리할 Controller 클래스를 찾은 뒤 실제로 클라이언트 요청을 처리할 Handler 메소드를 찾아 호출하는데, DispatcherServlet은 Handler 메소드를 직접 호출하지 않고 HandlerAdapter에게 Handler 메소드 호출을 위임함
+5. HandlerAdapter는 DispatcherServlet으로부터 전달받은 Controller 정보를 기반으로 해당 Controller의 Handler 메소드를 호출
+6. Controller의 Handler 메소드는 비즈니스 로직 처리 후 리턴 받은 Model 데이터를 HandlerAdapter에게 전달
+7. HandlerAdapter는 전달받은 Model 데이터와 View 정ㅂ를 ㄷ시 DispatcherServlet에게 전달 (HandlerAdapter는 어떤 리턴 타입이든 모두 ModelAndView로 반환함)
+8. DispatcherServlet은 전달받은 View 정보를 다시 ViewResolver에게 전달하여 View 검색 요청
+9. ViewResolver는 View 정보에 해당하는 View를 찾아서 View를 다시 리턴함
+10. DispatcherServlet은 ViewResolver로부터 전달받은 View 객체를 통해 Model 데이터를 넘겨주면서 클라이언트에게 전달할 응답 데이터 생성을 요청
+11. View는 응답 데이터를 생성해서 다시 DispatcerServlet에게 전달
+12. DispatcherServlet은 전달받은 응답 데이터를 최종적으로 클라이언트에게 전달
```
이걸 이해하셨으면 spring고수가 될 수 있을겁니다ㅎㅎㅎ

enohs replied:
ㅋㅋㅋㅋ이해하고 싶네요ㅠ 따로 더 알아보겠습니다😁

KoSeonJe commented on week06/[6주차]김시온.md:
```diff
@@ -0,0 +1,228 @@
+   
+그 다음에는 간략히 나타내면, 어노테이션 처리를 활성화하고, 코드에 어노테이션을 추가한 뒤 응용프로그램을 빌드하고 실행하면 된다.
+   
+어노테이션을 사용할 때 알아두면 좋을 것들이 있다. 일단 어노테이션을 잘 사용하면 강력한 도구가 될 수 있지만 남용을 할 경우 읽기 어려워질 수 있고, 불필요한 종속성과 결합을 만들 수 있다. 어노테이션이 달린 클래슨 메소드에 설명 이름을 사용하는 것이 구성 요소의 목적과 동작을 더 쉽게 이해하는 데 도움을 줄 수 있다. 어노테이션 특성에 표준 명명 규칙을 사용하는 것이 좋다. 사용자 지정 어노테이션을 정의하거나 속성을 사용할 때는 속성 이름에 카멜표기법을 사용하고 단일 값 속석에 값으로 접두사를 붙이는 등의 규칙을 따르는 것이 좋다. 이 밖에도 특정 사용 사례에 적합한 어노테이션을 사용하고, 구성요소 검색에 스프링 스테리오 타입 어노테이션 사용, 어노테이션 메타 데이터를 체계적으로 유지, 어노테이션의 사 ...
```
내용 좋습니다

KoSeonJe commented on week06/[6주차]김시온.md:
```diff
@@ -0,0 +1,228 @@
+
+# 선택
+
+## Spring Framework 구조와 동작 방식
```
Filter, Interceptor, Converter, Argument Resolver 의 키워드도 조사해보면 좋을 것 같습니다

enohs replied:
아니... 선제 멘토님 정성 미쳤습니다......
피드백 부분 참고해서 더 공부해볼게요~!

KoSeonJe reviewed (commented):
열심히 하셨네요! 고생하셨습니다

enohs committed:
chore : 초기 문서 설정

enohs committed:
Merge branch 'COW-edu:main' into main

enohs committed:
Merge remote-tracking branch 'origin/main'

enohs committed:
feat : 멤버 회원가입 구현

enohs committed:
Merge branch 'COW-edu:main' into main

enohs committed:
feat : 멤버 회원가입 구현 수정

enohs committed:
feat : 멤버 찾기 구현
feat : 멤버 수정 구현
feat : 멤버 삭제 구현
feat : 포스트 작성 구현
feat : 포스트 모두 찾기 구현
feat : 포스트 일부 찾기 구현
feat : 포스트 수정 구현
feat : 포스트 삭제 구현

KoSeonJe commented on week07/kimsion-crud-api/src/main/java/practice/kimsioncrudapi/member/CreateMemberRequest.java:
```diff
@@ -0,0 +1,21 @@
+
+  public Member toEntity(){
+    return Member.builder()
+        .password(password) //
+        .name(username) //
+        .email(email) //
```
주석 표시 제거해도 좋을 것 같습니다.

enohs replied:
사실 저 주석 표시가 없으면 자동으로 포맷팅 되면서 일렬로만 써지더라고요ㅠ 
임시방편으로 엔터키 입력을 위해 사용했습니다만 원인을 찾아서 변경해보겠습니다

KoSeonJe commented on week07/kimsion-crud-api/src/main/java/practice/kimsioncrudapi/member/Member.java:
```diff
@@ -0,0 +1,27 @@
+  private String name;
+  private String email;
+  private String password;
+  private Integer id;
```
Integer로 선언한 이유가 있나요?
Integer와 int의 차이를 알아보면 좋을 것 같습니다.
Integer가 메모리를 더 차지하기 때문에, 컬렉션을 사용한는 경우 혹은 null체크 하고싶은 것이 아니라면 int를 사용하면 좋을 것 같습니다.

enohs replied:
Mapping하면서 인자로 Integer를 써야해서 수정하다가 전부 Integer로 바꿨습니다😅 그런 디테일은 또 몰랐네요...
참고하겠습니다

KoSeonJe commented on week07/kimsion-crud-api/src/main/java/practice/kimsioncrudapi/member/Member.java:
```diff
@@ -0,0 +1,27 @@
+  private String password;
+  private Integer id;
+
+  @Builder // new 키워드
```
Builder패턴을 정확히 아시고 사용하시면 좋을 것 같으니 찾아보시면 좋을 것 같아요!

KoSeonJe commented on week07/kimsion-crud-api/src/main/java/practice/kimsioncrudapi/member/Member.java:
```diff
@@ -0,0 +1,27 @@
+    this.name = createMemberRequest.getUsername();
+    this.email = createMemberRequest.getEmail();
+    this.password = createMemberRequest.getPassword();
+  }
```
엔터티 내에 DTO를 의존하고 있는데, 괜찮은 방법일까요? 스스로 찾아보시고 고민해보면 좋을 것 같아요

Hoya324 commented on week07/kimsion-crud-api/src/main/java/practice/kimsioncrudapi/member/CreateMemberRequest.java:
```diff
@@ -0,0 +1,21 @@
+import lombok.Getter;
+
+@Getter
+public class CreateMemberRequest { // DTO
+
+  private String username;
+  private String email;
+  private String password;
+  private Integer id;
+
+  public Member toEntity(){
+    return Member.builder()
+        .password(password) //
+        .name(username) //
+        .email(email) //
+        .id(id)
+        .build();
```
주석을 다시면서 공부하시는거라면 좋지만 불필요한 주석은 지우는 습관이 필요합니다!

KoSeonJe commented on week07/kimsion-crud-api/src/main/java/practice/kimsioncrudapi/member/MemberController.java:
```diff
@@ -0,0 +1,38 @@
+  }
+
+  @PatchMapping("/members/{id}")
+  public MemberResponse patchMember(@PathVariable("id") Integer id, @RequestBody CreateMemberRequest createMemberRequest){
+    return memberService.updateInformation(id, createMemberRequest);
```
생성과 수정을 같은 dto를 사용하고 계신데, 코드의 중복을 피하시고 싶어가지고 이런 방법을 선택한것으로 보입니다.
하지만, 생성과 수정등 API의 DTO는 추후에  변하기 쉽기 때문에 각자 자신의 dto를 생성해주시는 것이 좋습니다.

enohs replied:
아 하나에 더 넣는 것보다는 나누는 것이 좋군요! 알겠습니다!!

KoSeonJe commented on week07/kimsion-crud-api/src/main/java/practice/kimsioncrudapi/member/MemberResponse.java:
```diff
@@ -0,0 +1,26 @@
+    this.id = id;
+  }
+
+  public static MemberResponse from(Member member){
+    return MemberResponse.builder().id(member.getId()).password(member.getPassword()).userName(member.getName()).email(member.getEmail()).build();
+  }
```
팩토리 메소드 생성 좋은 것 같아요.!
근데 builder를 통해 생성되는 인자의 가독성을 높이기 위해 줄바꿈 해주시면 좋을 것 같습니다

enohs replied:
줄바꿈 적용이 안돼서 설정에서 찾아보겠습니닷

Hoya324 commented on week07/kimsion-crud-api/src/main/java/practice/kimsioncrudapi/post/PostRepository.java:
```diff
@@ -0,0 +1,36 @@
+public class PostRepository {
+
+  private final Map<Integer, Post> postRepository = new ConcurrentHashMap<>();
+  int id = 0;
```
이렇게 default로 접근제어자가 설정되어 같은 패키지에서 접근이 가능하게 됩니다.!
id에 대해 다른 곳에서 접근하는 것은 치명적일 수 있으니 고려해보세요!!

enohs replied:
와 이런 실수를... 전혀 인지 못하고 있었네요😅 넵! 알겠습니다

Hoya324 commented on week07/kimsion-crud-api/src/main/java/practice/kimsioncrudapi/post/PostService.java:
```diff
@@ -0,0 +1,38 @@
+
+  private final PostRepository postRepository;
+
+  public void writePost(CreatePostRequest createPostRequest) {
```
write는 관점이 유저인거 같아요!
하지만 PostService가 게시물을 작성하진 않으니 생성의 관점으로 create를 쓰는게 저는 좋을거 같습니다.!
물론 사람마다 의견이 다를 순 있으니 그렇구나~해주세요!

Hoya324 reviewed (commented):
전반적으로 짜임새가 좋은 crud 같습니다!
너무너무 수고하셨어요!

KoSeonJe commented on week07/kimsion-crud-api/src/main/java/practice/kimsioncrudapi/post/PostService.java:
```diff
@@ -0,0 +1,38 @@
+  public PostResponse updatePost(Integer id, CreatePostRequest createPostRequest) {
+    Post post = postRepository.findById(id);
+    post.update(createPostRequest);
+    postRepository.update(post);
```
PostRepository에서 읽어온 Post를 직접 update를 통해 변경해주었기 때문에, 해당 메소드는 필요 없을 것 같습니다.
한번 테스트 해보시면 좋을 것 같아요

enohs replied:
아하 그러면 업데이트하고 레포지토리에 다시 저장할 필요는 없는 건가요? 직접 테스트해봐야 할 것 같네요ㅎㅎ

KoSeonJe replied:
id를 사용하여 저장된 Post 객체를 찾아온다면, 저장되어있는 객체와 찾아온 객체가 주소가 동일하기 때문에 자동을로 값이 변경됩니다.
고민해보고 이해하시면 좋을 것 같아요!

KoSeonJe reviewed (commented):
고생하셨습니다! 정말 잘하시고 있는 것 같아요!

enohs committed:
refactor : JPA 적용

enohs committed:
docs : md 파일 추가

enohs committed:
Update [7주차]김시온.md

docs : 오탈자 수정

enohs committed:
refactor : cascade 속성 추가

enohs committed:
docs : md파일 수정
